File: frontend/src/api/runs.ts
import { apiFetch } from "./client";
import type { Run } from "../types/run";

export const listRuns = (projectId: string, taskId: string) =>
  apiFetch<Run[]>(
    `/projects/${encodeURIComponent(projectId)}/tasks/${encodeURIComponent(taskId)}/runs`,
  );

export const getRun = (projectId: string, taskId: string, runId: string) =>
  apiFetch<Run>(
    `/projects/${encodeURIComponent(projectId)}/tasks/${encodeURIComponent(taskId)}/runs/${encodeURIComponent(runId)}`,
  );

File: backend/src/handlers/runs.rs
use std::sync::Arc;

use axum::{
    Json,
    extract::{Path, State},
    response::IntoResponse,
};

use crate::error::AppError;
use crate::services;
use crate::state::AppState;

pub async fn list_runs(
    State(state): State<Arc<AppState>>,
    Path((project_id, task_id)): Path<(String, String)>,
) -> Result<impl IntoResponse, AppError> {
    let runs = services::runs::list_runs_for_task(&state.db, &project_id, &task_id).await?;
    Ok(Json(runs))
}

pub async fn get_run(
    State(state): State<Arc<AppState>>,
    Path((project_id, task_id, run_id)): Path<(String, String, String)>,
) -> Result<impl IntoResponse, AppError> {
    let run = services::runs::get_run_by_id(&state.db, &project_id, &task_id, &run_id).await?;
    Ok(Json(run))
}
